src/QsCompiler/Transformations/CallGraph.cs

```diff
@@ -0,0 +1,569 @@
+
+            foreach (var resolutionDictionary in resolutionDictionaries)
+            {
+                // Contains a lookup of all the keys in the combined resolutions whose value needs to be updated
+                // if a certain type parameter is resolved by the currently processed dictionary.
+                var mayBeReplaced = combinedBuilder
+                    .Where(kv => kv.Value.Resolution.IsTypeParameter)
+                    .ToLookup(
+                        kv => AsTypeResolutionKey(((ResolvedTypeKind.TypeParameter)kv.Value.Resolution).Item),
+                        entry => entry.Key);
+
+                // Do any replacements for type parameters that may be replaced with values in the current dictionary.
+                // This needs to be done first to cover an edge case.
+                foreach (var (typeParam, paramRes) in resolutionDictionary.Where(entry => mayBeReplaced.Contains(entry.Key)))
+                {
+                    // Get all the parameters whose value is dependent on the current resolution's type parameter,
+                    // and update their values with this resolution's value.
+                    foreach (var keyInCombined in mayBeReplaced[typeParam])
+                    {
+                        // Check that we are not doing an excessively constrictive resolution.
+                        success = success && !IsConstrictiveResolution(keyInCombined, paramRes);
+                        combinedBuilder[keyInCombined] = paramRes;
+                    }
+                }
+
+                // Add all resolutions to the current dictionary.
+                foreach (var (typeParam, paramRes) in resolutionDictionary)
+                {
+                    // Check that we are not doing an excessively constrictive resolution.
+                    success = success && !IsConstrictiveResolution(typeParam, paramRes);
+                    // Check that there is no conflicting resolution already defined.
+                    var conflictingResolutionExists = combinedBuilder.TryGetValue(typeParam, out var current)
+                        && !current.Equals(paramRes) && !IsSelfResolution(typeParam, current);
+                    success = success && !conflictingResolutionExists;
+                    combinedBuilder[typeParam] = paramRes;
+                }
+            }
+
+            combined = combinedBuilder.ToImmutable();
+            return success;
+        }
+    }
+
+    /// <summary>
+    /// Struct containing information that exists on edges in a call graph.
+    /// </summary>
+    public struct CallGraphEdge
+    {
+        public TypeParameterResolutions ParamResolutions;
+    }
+
+    /// <summary>
+    /// Struct containing information that exists on nodes in a call graph.
+    /// </summary>
+    public struct CallGraphNode
```

Comment:
At least for CallGraphNode you have public methods below that expect members to be non-null. For that reason and also to align with how the rest of the code base works I would suggest to make this a class (and also CallGraphEdge) with a constructor that enforces that none of the members is ever null. Alternatively, you could look into whether it make sense to use C#'s nullable feature for that purpose.

Reply:
The reason they are structs is so that they work much better as dictionary keys then if they were classes, which is their primary use. I am not seeing where the members of this are expected to be non-null. Can you please point these out to me?

Reply:
I see. RemovePositionFromEdge and CombineEdges (via TryCombineTypeResolutions). You don't gain anything by having CallGraphEdge be a struct, since the ParamResolutions don't support support structural equality in any case.

Reply:
I've converted these to classes, added the constructors that require non-null arguments, and added non-null checks to the public methods.